File: kernel/transport/serve.mjs
// Kernel shim: turn a JavaScript module into a process that speaks the wire.
//
// This exists so `run.js` stays pure. The agent writes functions; it does not
// write plumbing, and it certainly does not write plumbing afresh in every
// module where a subtle difference could become a behaviour difference. See
// PROTOCOL.md for the format.
//
// Trusted kernel code. Never delegated.

import { createInterface } from "node:readline";
import { pathToFileURL } from "node:url";
import { resolve } from "node:path";

const entry = process.argv[2] ?? "run.js";
const name = process.argv[3] ?? "module";

let impl;
try {
  impl = await import(pathToFileURL(resolve(entry)).href);
} catch (err) {
  // A module that will not even load is not a protocol error, it is a dead
  // module — so this is one of the few places exiting is right.
  process.stderr.write(`serve: cannot load ${entry}: ${err instanceof Error ? err.stack : String(err)}\n`);
  process.exit(70);
}

const operations = Object.keys(impl).filter((k) => typeof impl[k] === "function").sort();

const rl = createInterface({ input: process.stdin, crlfDelay: Infinity });

for await (const line of rl) {
  const text = line.trim();
  if (text === "") continue;

  let req;
  try {
    req = JSON.parse(text);
  } catch {
    write({ id: null, error: { code: "EPROTOCOL", message: "request was not valid JSON" } });
    continue;
  }

  const id = req.id ?? null;

  if (req.op === "__describe") {
    write({ id, out: { module: name, operations } });
    continue;
  }

  const fn = typeof req.op === "string" ? impl[req.op] : undefined;
  if (typeof fn !== "function") {
    write({ id, error: { code: "ENOOP", message: `no operation ${JSON.stringify(req.op)}; this module exposes ${operations.join(", ") || "nothing"}` } });
    continue;
  }

  try {
    const out = await fn(req.in);
    write({ id, out });
  } catch (err) {
    // A declared error is a RESPONSE, not a crash. The process stays alive,
    // because "this input was rejected" and "this module is broken" are
    // different facts and a caller has to be able to tell them apart.
    write({
      id,
      error: {
        code: /** @type {any} */ (err)?.code ?? "EUNCAUGHT",
        message: err instanceof Error ? err.message : String(err),
      },
    });
  }
}

/** @param {unknown} obj */
function write(obj) {
  // stdout carries responses and nothing else. Anything a module wants to say to
  // a human belongs on stderr; a stray line here corrupts the stream.
  process.stdout.write(JSON.stringify(obj) + "\n");
}

File: kernel/transport/serve.py
# Kernel shim: turn a Python module into a process that speaks the wire.
#
# The exact counterpart of serve.mjs, and it must behave identically — a shim
# that is subtly more forgiving than its sibling would let a module in one
# language be admitted on weaker evidence than the same module in another.
# See PROTOCOL.md.
#
# Trusted kernel code. Never delegated.

import importlib.util
import inspect
import json
import os
import sys


def _load(entry):
    spec = importlib.util.spec_from_file_location("module_under_test", os.path.abspath(entry))
    if spec is None or spec.loader is None:
        raise ImportError(f"cannot load {entry}")
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


def _write(obj):
    # stdout carries responses and nothing else. Anything a module wants to say
    # to a human belongs on stderr; a stray print() here corrupts the stream.
    sys.stdout.write(json.dumps(obj) + "\n")
    sys.stdout.flush()


def main():
    entry = sys.argv[1] if len(sys.argv) > 1 else "run.py"
    name = sys.argv[2] if len(sys.argv) > 2 else "module"

    try:
        impl = _load(entry)
    except Exception as err:                      # noqa: BLE001 — a dead module is a dead module
        # A module that will not even load is not a protocol error. This is one
        # of the few places exiting is right.
        sys.stderr.write(f"serve: cannot load {entry}: {err!r}\n")
        sys.exit(70)

    operations = sorted(
        n for n, v in vars(impl).items()
        if not n.startswith("_") and (inspect.isfunction(v) or inspect.isbuiltin(v))
        and getattr(v, "__module__", None) == impl.__name__
    )

    for line in sys.stdin:
        text = line.strip()
        if not text:
            continue

        try:
            req = json.loads(text)
        except json.JSONDecodeError:
            _write({"id": None, "error": {"code": "EPROTOCOL", "message": "request was not valid JSON"}})
            continue

        rid = req.get("id")

        if req.get("op") == "__describe":
            _write({"id": rid, "out": {"module": name, "operations": operations}})
            continue

        op = req.get("op")
        fn = getattr(impl, op, None) if isinstance(op, str) else None
        if not callable(fn) or op not in operations:
            exposed = ", ".join(operations) or "nothing"
            _write({"id": rid, "error": {"code": "ENOOP", "message": f"no operation {json.dumps(op)}; this module exposes {exposed}"}})
            continue

        try:
            out = fn(req.get("in"))
            _write({"id": rid, "out": out})
        except Exception as err:                  # noqa: BLE001
            # A declared error is a RESPONSE, not a crash. The process stays
            # alive, because "this input was rejected" and "this module is
            # broken" are different facts a caller must be able to tell apart.
            code = getattr(err, "code", None) or "EUNCAUGHT"
            _write({"id": rid, "error": {"code": code, "message": str(err)}})


if __name__ == "__main__":
    main()
